torch/csrc/jit/tensorexpr/kernel.cpp

```diff
@@ -2487,6 +2488,78 @@ void fuseAllLoops(StmtPtr st) {
   }
 }
 
+// Prune innermost loops until iterations satisfies a minimum grain size.
+static void pruneByGrainSize(std::vector<ForPtr>& loops) {
+  constexpr int64_t minGrainSize = 32768;
+  int64_t grainSize = 1;
+  for (int64_t i = loops.size(); i > 0; i--) {
+    auto const& loop = loops[i - 1];
+    if (auto stop = to<IntImm>(loop->stop())) {
+      grainSize *= stop->value();
+      if (grainSize < minGrainSize) {
+        loops.pop_back();
+      }
+      continue;
+    }
+    break;
+  }
+}
+
+// Retain enough outermost loops to fill the number of threads.
+static void pruneByThreadCount(std::vector<ForPtr>& loops) {
+  int64_t trips = 1;
+  auto threads = at::get_num_threads();
+  auto it = loops.begin();
+  for (; it != loops.end(); it++) {
+    if (trips >= threads) {
+      break;
+    }
+    auto const& loop = *it;
+    if (auto stop = to<IntImm>(loop->stop())) {
+      trips *= stop->value();
+      continue;
+    }
+    break;
+  }
+  loops.erase(it, loops.end());
+}
+
+// Flatten and parallelize outer loops, subject to a minimum number of elements
+// in the inner loop, and a maximum level of thread-level parallelism in the
+// outer loops.
+template <typename Bufs>
+static void parallelizeOuterLoops(LoopNest& l, Bufs&& bufs) {
+  for (auto const& buf : bufs) {
+    auto loops = l.getLoopStmtsFor(buf);
```

Comment:
Since this function is called after `fuseAllLoops` it is possible that multiple buffers belong to the same loopnest. So, we could be repeating this loop multiple times for the same loopnest. I understand that that may not be incorrect at this point, but it could lead to bugs in future.

IMO, we shouldn't be looking at output buffers and their loopnests. Instead, we should just take `root_stmt` in the given `LoopNest` and apply parallelization for all loopnests in that stmt. Wdyt?

Reply:
That seems OK to me, sure.

Reply:
I guess where things could get a little weird is if multiple buffers are updated at different levels of the loopnest, e.g.:
```
for i:
  y1[] = ...
  for j:
    y2[] = ...
```

The current approach sort of gives each buffer an "independent" chance to affect the loop parallelization.  That doesn't seem terrible, tbh.

Reply:
Actually the more I think about it, is there really any advantage to starting with the root stmt and working down?  From my POV it just makes the code a lot more complicated; the way things happen now I just get a nice vector of loops leading to a buffer and try to flatten them.  If it's not flattenable, it simply fails and I give up.

Although, maybe it's not too much work to build up my own vector starting from the root.  Idk.

Reply:
If you are okay with having the same set of loops being handled here for different bufs, then I have no objections to it. 

Personally, I felt starting from the root_stmt might be better. We might need another API to extract all loops in the root_stmt. So, may be we can do this in future.

Reply:
If I understand correctly,  going through all buffers would not miss parallelism opportunities like the following
```
for i 
  for j1
      y1 = ... [data dependence exists between iterations]
  for j2
     y2 = ... [no data dependence between iterations]
``` 
`i+j1` cannot be parallelized because there's data dependence between iterations for `y1`; but `i+j2` can be parallelized and we should not miss it. If this is the thing we try to do here, I guess we need a `distribute` transformation before `flatten`. `flatten` currently only handles perfectly nested loops.

Reply:
Yeah the approach here will definitely miss opportunities where some nested loops are parallelizable.  It's really kind of a best-effort thing to get simple elementwise fusions right, not a general solution to parallelism.